# DS-SF-36 | 03 | `pandas` | Assignment | Starter Code

## `pandas` and Bistro

You've just told one of your friend that you are taking a Data Science class.  (Yeah!)  Your friend is running a bistro, a small restaurant, serving moderately priced simple meals in a modest setting ([Wikipedia](https://en.wikipedia.org/wiki/Bistro)).  She collected over some period of time the following information of her patrons' visits.

| Variable's name | Its meaning |
|:---:|:---|
| `name` | Patron's first name |
| `gender` | Patron's gender |
| `is_smoker` | Whether the patron is smoking or not |
| `party` | Party's size |
| `check` | Check amount (\$) (after taxes but before tips) |
| `tip` | Tip (\$) that the patron added to the check |
| `day` | Week day of the visit |
| `time` | Rough time estimate of the visit |

In this assignment, we will be exploring this dataset using `pandas`.<sup>(*)</sup>

<sup>(*)</sup> this dataset was adapted from the `tips` dataset of the `seaborn` package (https://github.com/mwaskom/seaborn-data)

> ### Question 1.  Import `pandas` (as `pd`).

In [3]:
import os
import pandas as pd
import numpy as np

# TODO

pd.set_option('display.max_rows', 10)
pd.set_option('display.notebook_repr_html', True)
pd.set_option('display.max_columns', 10)

> ### Question 2.  Read the `dataset-03-bistro.csv` dataset.

In [4]:
# TODO
!pwd

/Users/nikka.keivanfar/DS-SF-36/classes/03/notebooks


In [5]:
!ls ../datasets

dataset-03-bistro.csv              dataset-03-zillow-transactions.csv
dataset-03-zillow-properties.csv


In [6]:
!ls

assignment-03-pandas-starter-code-RyanK practice renaming.ipynb
assignment-03-pandas-starter-code.ipynb
codealong-03-pandas-starter-code-RyanK practice.ipynb
codealong-03-pandas-starter-code.ipynb


In [7]:
!cat ../datasets/dataset-03-bistro.csv ###cat = show me the contents of this file

In [8]:
df = pd.read_csv(os.path.join('..', 'datasets', 'dataset-03-bistro.csv'))  ###assign it to a data frame called "df"

> ### Question 3.  What is the class of the `pandas` object storing the dataset?

In [9]:
type(df)

pandas.core.frame.DataFrame

Answer: a dataframe


> ### Question 4.  How many samples (i.e., rows) are in this dataset?

In [10]:
df

day    time      name  gender  is_smoker  party  check   tip
0      Sunday  Dinner  Kimberly  Female      False      2  16.99  1.01
1      Sunday  Dinner  Nicholas    Male      False      3  10.34  1.66
2      Sunday  Dinner     Larry    Male      False      3  21.01  3.50
3      Sunday  Dinner    Joseph    Male      False      2  23.68  3.31
4      Sunday  Dinner    Janice  Female      False      4  24.59  3.61
..        ...     ...       ...     ...        ...    ...    ...   ...
239  Saturday  Dinner     Kevin    Male      False      3  29.03  5.92
240  Saturday  Dinner    Sandra  Female      False      2  27.18  2.00
241  Saturday  Dinner      Carl    Male      False      2  22.67  2.00
242  Saturday  Dinner       Jon    Male      False      2  17.82  1.75
243  Thursday  Dinner    Brandi  Female      False      2  18.78  3.00

[244 rows x 8 columns]

In [11]:
len(df)

244

Answer: 244 rows

> ### Question 5.  How many variables (i.e., columns) are in this dataset?

In [ ]:
# TODO

Answer: 8 columns


> ### Question 6.  Print the name of each column in the dataset, one name per line.

In [12]:
list(df)

['day', 'time', 'name', 'gender', 'is_smoker', 'party', 'check', 'tip']

In [14]:
df.columns.tolist()

['day', 'time', 'name', 'gender', 'is_smoker', 'party', 'check', 'tip']

In [15]:
df.columns

Index([u'day', u'time', u'name', u'gender', u'is_smoker', u'party', u'check',
       u'tip'],
      dtype='object')

> ### Question 7.  Print the first two rows of the dataset to the console.  What does the output look like?

In [17]:
df.head(2)

day    time      name  gender  is_smoker  party  check   tip
0  Sunday  Dinner  Kimberly  Female      False      2  16.99  1.01
1  Sunday  Dinner  Nicholas    Male      False      3  10.34  1.66

Answer: output is a dataframe 

> ### Question 8.  Extract the last 2 rows of the data frame and print them to the console.  What does the output look like?

In [18]:
df.tail(2)

day    time    name  gender  is_smoker  party  check   tip
242  Saturday  Dinner     Jon    Male      False      2  17.82  1.75
243  Thursday  Dinner  Brandi  Female      False      2  18.78  3.00

Answer: output is a dataframe

> ### Question 9.  Does the dataset contain any missing values?

In [19]:
df.isnull()

day   time   name  gender  is_smoker  party  check    tip
0    False  False  False   False      False  False  False  False
1    False  False  False   False      False  False  False  False
2    False  False  False   False      False  False  False  False
3    False  False  False   False      False  False  False  False
4    False  False  False   False      False  False  False  False
..     ...    ...    ...     ...        ...    ...    ...    ...
239  False  False  False   False      False  False  False  False
240  False  False  False   False      False  False  False  False
241  False  False  False   False      False  False  False  False
242  False  False  False   False      False  False  False  False
243  False  False  False   False      False  False  False  False

[244 rows x 8 columns]

In [20]:
df.isnull().sum()

day          0
time         0
name         0
gender       0
is_smoker    0
party        0
check        0
tip          0
dtype: int64

In [21]:
df.isnull().sum().sum()  ###tells you number of cell with missing values

0L

Answer: no missing values

> ### Question 10.  What can you say about the `is_smoker` variable?  I.e., will it bring any insights when analyzing the dataset?  What do you want to do with it?  (and do it...)

In [22]:
df['is_smoker']

0      False
1      False
2      False
3      False
4      False
       ...  
239    False
240    False
241    False
242    False
243    False
Name: is_smoker, Length: 244, dtype: bool

In [23]:
df = df.set_index('is_smoker')  ###i'm not sure this did what i wanted it to do...

In [25]:
df.index

Index([False, False, False, False, False, False, False, False, False, False,
       ...
       False, False, False, False, False, False, False, False, False, False],
      dtype='object', name=u'is_smoker', length=244)

In [26]:
df

day    time      name  gender  party  check   tip
is_smoker                                                        
False        Sunday  Dinner  Kimberly  Female      2  16.99  1.01
False        Sunday  Dinner  Nicholas    Male      3  10.34  1.66
False        Sunday  Dinner     Larry    Male      3  21.01  3.50
False        Sunday  Dinner    Joseph    Male      2  23.68  3.31
False        Sunday  Dinner    Janice  Female      4  24.59  3.61
...             ...     ...       ...     ...    ...    ...   ...
False      Saturday  Dinner     Kevin    Male      3  29.03  5.92
False      Saturday  Dinner    Sandra  Female      2  27.18  2.00
False      Saturday  Dinner      Carl    Male      2  22.67  2.00
False      Saturday  Dinner       Jon    Male      2  17.82  1.75
False      Thursday  Dinner    Brandi  Female      2  18.78  3.00

[244 rows x 7 columns]

Answer: TODO

> ### Question 11.  For which week days does the dataset has data for?

In [27]:
df.day

is_smoker
False      Sunday
False      Sunday
False      Sunday
False      Sunday
False      Sunday
           ...   
False    Saturday
False    Saturday
False    Saturday
False    Saturday
False    Thursday
Name: day, Length: 244, dtype: object

In [31]:
df.day.unique()

array(['Sunday', 'Saturday', 'Thursday', 'Friday'], dtype=object)

Answer: Thursday and Friday

> ### Question 12.  How often was the bistro patronized for each week day?

(check `.value_counts()`; it could come in handy)

(http://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.value_counts.html)

In [32]:
df[['day', 'party']]

day  party
0      Sunday      2
1      Sunday      3
2      Sunday      3
3      Sunday      2
4      Sunday      4
..        ...    ...
239  Saturday      3
240  Saturday      2
241  Saturday      2
242  Saturday      2
243  Thursday      2

[244 rows x 2 columns]

In [38]:
df['day'].value_counts()

Saturday    87
Sunday      76
Thursday    62
Friday      19
Name: day, dtype: int64

Answer: Thursday = 62, Friday = 19

> ### Question 13.  How much tip did waiters collect for each week day?

In [39]:
df['tip'].value_counts()

2.00    33
3.00    23
4.00    12
5.00    10
2.50    10
        ..
2.83     1
1.58     1
3.71     1
3.35     1
2.18     1
Name: tip, Length: 123, dtype: int64

In [42]:
list(set(df.day)) ###create a list of unique values by turning the pandas column into a set

['Sunday', 'Friday', 'Saturday', 'Thursday']

In [49]:
thursday = df[df["day"]=="Thursday"]
friday = df[df["day"]=="Friday"]
print(thursday["tip"]).sum()
print(friday["tip"]).sum()

171.83
51.96


Answer: thursday = $171.83
        friday = $51.96

> ### Question 14.  What is the average tip per check (in absolute \$) for each week day?

In [50]:
print(thursday["tip"]).mean()
print(friday["tip"]).mean()

2.7714516129
2.73473684211


Answer: thursday = $2.77
        friday = $2.73

> ### Question 15.  What is the average tip per check (as a percentage of the check) for each week day?

In [68]:
thurs_tip_avg_percent = (thursday['tip']/thursday['check']*100).mean()
fri_tip_avg_percent = (friday['tip']/friday['check']*100).mean()

thurs_tip_avg_percent




16.127563396664712

In [69]:
fri_tip_avg_percent

16.991302873347887

Answer: thurs = 16.13%
        fri = 16.99%

> ### Question 16.  Are there any name in common between male and female patrons?  (E.g., `Chris` can refer to either a man or a woman)

(check `numpy.intersect1d()`; it could come in handy)

(https://docs.scipy.org/doc/numpy/reference/generated/numpy.intersect1d.html)

In [74]:
df

is_smoker       day    time      name  gender  party  check   tip
0        False    Sunday  Dinner  Kimberly  Female      2  16.99  1.01
1        False    Sunday  Dinner  Nicholas    Male      3  10.34  1.66
2        False    Sunday  Dinner     Larry    Male      3  21.01  3.50
3        False    Sunday  Dinner    Joseph    Male      2  23.68  3.31
4        False    Sunday  Dinner    Janice  Female      4  24.59  3.61
..         ...       ...     ...       ...     ...    ...    ...   ...
239      False  Saturday  Dinner     Kevin    Male      3  29.03  5.92
240      False  Saturday  Dinner    Sandra  Female      2  27.18  2.00
241      False  Saturday  Dinner      Carl    Male      2  22.67  2.00
242      False  Saturday  Dinner       Jon    Male      2  17.82  1.75
243      False  Thursday  Dinner    Brandi  Female      2  18.78  3.00

[244 rows x 8 columns]

Answer: TODO

> ### Question 17.  If no patrons share the same name, how many unique patrons are in the dataset?

In [ ]:
# TODO

Answer: TODO

> ### Question 18.  How many times did `Kevin` patronized the bistro?  How about `Alice`?

In [ ]:
# TODO

Answer: TODO

> ### Question 19.  Who are the top 3 female and male patrons?

In [ ]:
# TODO

Answer: TODO

> ### Question 20.  Who's the best tipper (as a fraction of all tips over all check totals)?  Who's the worst?  How many times did they patronize the bistro?

- (http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.sort_values.html)
- (http://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.sort_values.html)

In [ ]:
# TODO

Answer: TODO